File: client/ts/src/teammate.ts
// teammate.ts — Teammate role: session join, task execution, explore, approvals.

import type { CloudEvent, ExploreRequest, Task, TeamClientConfig } from './types.js';
import {
  EventApprovalRequest,
  EventExploreRequest,
  EventTaskAssigned,
  EventTaskClaim,
  EventTaskComplete,
  EventTaskFail,
  EventTaskProgress,
} from './types.js';
import type { ApprovalPlugin, ExplorePlugin, ProgressReporter, TeammatePlugin } from './plugin.js';
import type { WSConnection } from './ws.js';
import { newEvent } from './client.js';

export interface TeammateAgentDeps {
  teammatePlugin: TeammatePlugin | null;
  approvalPlugin: ApprovalPlugin | null;
  explorePlugin: ExplorePlugin | null;
  doJSON: <T>(method: string, path: string, body?: unknown) => Promise<T>;
}

/** TeammateAgent owns all teammate-role logic and handles relevant inbound events. */
export class TeammateAgent {
  private cfg: TeamClientConfig;
  private ws: WSConnection;
  private deps: TeammateAgentDeps;
  private memberId = '';

  constructor(cfg: TeamClientConfig, ws: WSConnection, deps: TeammateAgentDeps) {
    this.cfg = cfg;
    this.ws = ws;
    this.deps = deps;
  }

  /** Registers this machine as a session member via REST. */
  async init(_signal: AbortSignal): Promise<void> {
    const resp = await this.deps.doJSON<{ id: string }>(
      'POST',
      `/api/team/sessions/${this.cfg.sessionId}/members`,
      {
        machineId: this.cfg.machineId,
        machineName: this.cfg.machineName ?? '',
      }
    );
    this.memberId = resp?.id ?? '';
  }

  /** Dispatches inbound events relevant to the teammate role. */
  handle(evt: CloudEvent): void {
    switch (evt.type) {
      case EventTaskAssigned:
        this.handleTaskAssigned(evt);
        break;

      case EventExploreRequest:
        this.handleExploreRequest(evt);
        break;
    }
  }

  /**
   * Sends an approval.request to the leader via WebSocket.
   * riskLevel should be "low", "medium", or "high".
   */
  requestApproval(
    toolName: string,
    description: string,
    riskLevel: string,
    toolInput: Record<string, unknown>
  ): void {
    this.ws.send(
      newEvent(EventApprovalRequest, this.cfg.sessionId, {
        toolName,
        description,
        riskLevel,
        toolInput,
      })
    );
  }

  /**
   * Registers a local repository with the session's affinity registry.
   * Call this on startup so the leader can schedule repo-specific tasks here.
   */
  async registerRepo(
    remoteUrl: string,
    localPath: string,
    branch: string,
    dirty: boolean
  ): Promise<void> {
    await this.deps.doJSON('POST', `/api/team/sessions/${this.cfg.sessionId}/repos`, {
      memberId: this.memberId,
      repoRemoteUrl: remoteUrl,
      repoLocalPath: localPath,
      currentBranch: branch,
      hasUncommittedChanges: dirty,
      lastSyncedAt: new Date().toISOString(),
    });
  }

  // ─── Private ─────────────────────────────────────────────────────────

  private handleTaskAssigned(evt: CloudEvent): void {
    if (!this.deps.teammatePlugin) return;

    const task = evt.payload?.['task'] as Task | undefined;
    if (!task?.id) return;

    // Claim immediately so the leader knows this task is being worked on.
    this.ws.send(
      newEvent(EventTaskClaim, this.cfg.sessionId, { taskId: task.id })
    );

    void this.executeTask(task);
  }

  private async executeTask(task: Task): Promise<void> {
    const ctrl = new AbortController();
    const reporter: ProgressReporter = {
      report: (pct, message) => {
        this.ws.send(
          newEvent(EventTaskProgress, this.cfg.sessionId, {
            taskId: task.id,
            percent: pct,
            message,
          })
        );
      },
    };

    // Signal start.
    this.ws.send(
      newEvent(EventTaskProgress, this.cfg.sessionId, {
        taskId: task.id,
        percent: 0,
        message: 'started',
      })
    );

    try {
      const result = await this.deps.teammatePlugin!.executeTask(
        ctrl.signal,
        task,
        reporter
      );
      this.ws.send(
        newEvent(EventTaskComplete, this.cfg.sessionId, {
          taskId: task.id,
          result,
        })
      );
    } catch (err) {
      this.ws.send(
        newEvent(EventTaskFail, this.cfg.sessionId, {
          taskId: task.id,
          errorMessage: err instanceof Error ? err.message : String(err),
        })
      );
    }
  }

  private handleExploreRequest(evt: CloudEvent): void {
    if (!this.deps.explorePlugin) return;

    const requestId = evt.payload?.['requestId'] as string | undefined;
    const fromMachineId = evt.payload?.['fromMachineId'] as string | undefined;
    if (!requestId) return;

    const req: ExploreRequest = {
      requestId,
      sessionId: this.cfg.sessionId,
      fromMachineId: fromMachineId ?? '',
      queries: (evt.payload?.['queries'] as ExploreRequest['queries']) ?? [],
    };

    void (async () => {
      const ctrl = new AbortController();
      try {
        const result = await this.deps.explorePlugin!.explore(ctrl.signal, req);
        this.ws.send(
          newEvent('explore.result', this.cfg.sessionId, {
            requestId: result.requestId,
            queryResults: result.queryResults,
            fromMachineId,
            error: result.error,
          })
        );
      } catch (err) {
        this.ws.send(
          newEvent('explore.result', this.cfg.sessionId, {
            requestId,
            queryResults: [],
            fromMachineId,
            error: err instanceof Error ? err.message : String(err),
          })
        );
      }
    })();
  }
}

File: client/go/team/teammate.go
package team

import (
	"context"
	"encoding/json"
	"fmt"
	"time"
)

// teammateAgent handles all teammate-role behaviour:
//   - joining the session (REST)
//   - executing assigned tasks via TeammatePlugin
//   - handling explore.request events via ExplorePlugin
//   - sending approval.request events to the leader
type teammateAgent struct {
	c        *Client
	memberID string // populated after init
}

func newTeammateAgent(c *Client) *teammateAgent {
	return &teammateAgent{c: c}
}

// init registers this machine as a session member via REST.
func (t *teammateAgent) init(_ context.Context) error {
	var resp struct {
		ID string `json:"id"`
	}
	err := t.c.doJSON("POST",
		"/api/team/sessions/"+t.c.cfg.SessionID+"/members",
		map[string]any{
			"machineId":   t.c.cfg.MachineID,
			"machineName": t.c.cfg.MachineName,
		},
		&resp,
	)
	if err != nil {
		return fmt.Errorf("join session: %w", err)
	}
	t.memberID = resp.ID
	return nil
}

// handle dispatches incoming server events relevant to the teammate role.
func (t *teammateAgent) handle(evt CloudEvent) {
	switch evt.Type {

	case EventTaskAssigned:
		taskRaw, _ := json.Marshal(evt.Payload["task"])
		var task Task
		if json.Unmarshal(taskRaw, &task) != nil || task.ID == "" {
			return
		}
		if t.c.teammatePlugin == nil {
			return
		}
		// Claim the task immediately so the leader knows it's being worked on.
		t.c.ws.send(newEvent(EventTaskClaim, t.c.cfg.SessionID, map[string]any{ //nolint:errcheck
			"taskId": task.ID,
		}))
		go t.executeTask(task)

	case EventExploreRequest:
		if t.c.explorePlugin == nil {
			return
		}
		requestID, _ := evt.Payload["requestId"].(string)
		if requestID == "" {
			return
		}
		go t.handleExplore(evt)

	case EventApprovalResponse:
		// The leader responded to our approval request. Optionally notify the plugin.
		if t.c.approvalPlugin == nil {
			return
		}
		// We receive only the response here, not a full ApprovalRequest, so we
		// build a minimal stub for the plugin in case it wants to log the outcome.
		approvalID, _ := evt.Payload["approvalId"].(string)
		status, _ := evt.Payload["status"].(string)
		feedback, _ := evt.Payload["feedback"].(string)
		if approvalID == "" {
			return
		}
		go func() {
			_ = approvalID
			_ = status
			_ = feedback
			// Host can extend via ApprovalPlugin if needed; no further action here.
		}()
	}
}

// executeTask runs the task through TeammatePlugin and reports progress/result.
func (t *teammateAgent) executeTask(task Task) {
	ctx := context.Background()
	reporter := &wsProgressReporter{c: t.c, taskID: task.ID}

	// Signal that we've started.
	t.c.ws.send(newEvent(EventTaskProgress, t.c.cfg.SessionID, map[string]any{ //nolint:errcheck
		"taskId":  task.ID,
		"percent": 0,
		"message": "started",
	}))

	result, err := t.c.teammatePlugin.ExecuteTask(ctx, task, reporter)
	if err != nil {
		t.c.ws.send(newEvent(EventTaskFail, t.c.cfg.SessionID, map[string]any{ //nolint:errcheck
			"taskId":       task.ID,
			"errorMessage": err.Error(),
		}))
		return
	}

	t.c.ws.send(newEvent(EventTaskComplete, t.c.cfg.SessionID, map[string]any{ //nolint:errcheck
		"taskId": task.ID,
		"result": result,
	}))
}

// handleExplore runs the explore request through ExplorePlugin and sends the result.
func (t *teammateAgent) handleExplore(evt CloudEvent) {
	requestID, _ := evt.Payload["requestId"].(string)
	fromMachineID, _ := evt.Payload["fromMachineId"].(string)

	// Re-marshal the payload into an ExploreRequest.
	raw, _ := json.Marshal(evt.Payload)
	var req ExploreRequest
	if err := json.Unmarshal(raw, &req); err != nil {
		req.RequestID = requestID
	}

	result, err := t.c.explorePlugin.Explore(context.Background(), req)
	if err != nil {
		result = ExploreResult{
			RequestID: requestID,
			Error:     err.Error(),
		}
	}
	result.RequestID = requestID

	t.c.ws.send(newEvent(EventExploreResult, t.c.cfg.SessionID, map[string]any{ //nolint:errcheck
		"requestId":     requestID,
		"queryResults":  result.QueryResults,
		"fromMachineId": fromMachineID,
		"error":         result.Error,
	}))
}

// RequestApproval sends an approval.request to the leader via WebSocket.
// riskLevel should be "low", "medium", or "high".
func (t *teammateAgent) RequestApproval(toolName, description, riskLevel string, toolInput map[string]any) error {
	return t.c.ws.send(newEvent(EventApprovalRequest, t.c.cfg.SessionID, map[string]any{
		"toolName":    toolName,
		"description": description,
		"riskLevel":   riskLevel,
		"toolInput":   toolInput,
	}))
}

// RegisterRepo registers a local repository with the session's affinity registry.
func (t *teammateAgent) RegisterRepo(remoteURL, localPath, branch string, dirty bool) error {
	return t.c.doJSON("POST",
		"/api/team/sessions/"+t.c.cfg.SessionID+"/repos",
		map[string]any{
			"memberId":              t.memberID,
			"repoRemoteUrl":         remoteURL,
			"repoLocalPath":         localPath,
			"currentBranch":         branch,
			"hasUncommittedChanges": dirty,
			"lastSyncedAt":          time.Now().Format(time.RFC3339),
		},
		nil,
	)
}

// ─── wsProgressReporter ───────────────────────────────────────────────────

// wsProgressReporter implements ProgressReporter by sending task.progress
// events over the WebSocket connection.
type wsProgressReporter struct {
	c      *Client
	taskID string
}

func (r *wsProgressReporter) Report(pct int, message string) {
	r.c.ws.send(newEvent(EventTaskProgress, r.c.cfg.SessionID, map[string]any{ //nolint:errcheck
		"taskId":  r.taskID,
		"percent": pct,
		"message": message,
	}))
}
